### PR TITLE
feat(metrics): Kill switch for transaction metric transaction [INGEST-333]

### PR DESCRIPTION
**Pull request by jjbayer**
Add a kill switch that sets the project config for transaction metrics extraction to an empty config. Once Relay refreshes its config for the given project, no more metrics will be extracted from transaction payloads.

To make sure that the project config is refreshed immediately, add a callback mechanism to kill switches which (for this kill switch) triggers an invalidation task for it.

Note: This mechanism might be still be too slow because it takes time for the project config to propagate to downstream Relays.

Requires option defined in https://github.com/getsentry/sentry/pull/36244.

### CODE REVIEW

File: src/sentry/killswitches.py [hunk 1]
```diff
@@ -129,6 +147,17 @@ class KillswitchInfo:
...
+    "relay.drop-transaction-metrics": KillswitchInfo(
+        description="""
+        Tell Relay via project config to stop extracting metrics from transactions.
+        Note that this change will not take effect immediately, it takes time
+        for downstream Relay instances to update their caches.
+        """,
+        fields={
+            "project_id": "project ID for which we want to stop extracting transaction metrics",
+        },
+        on_change=_update_project_configs,
```
Comment (jjbayer) [inline src/sentry/killswitches.py:159]: I tested this manually, not sure if I need another test in `test_killswitches.py`.

File: src/sentry/options/defaults.py [hunk 2]
```diff
@@ -362,6 +362,9 @@
 # All Relay options (statically authenticated Relays can be registered here)
 register("relay.static_auth", default={}, flags=FLAG_NOSTORE)
 
+# Tell Relay to stop extracting metrics from transaction payloads
+register("relay.drop-transaction-metrics", default=[])
```
Comment (jjbayer) [inline src/sentry/options/defaults.py:366]: Do I need another PR in `getsentry` for this?

File: src/sentry/relay/config/__init__.py [hunk 3]
```diff
@@ -417,9 +417,26 @@ def _filter_option_to_config_setting(flt, setting):
...
 def get_transaction_metrics_settings(
     project: Project, breakdowns_config: Optional[Mapping[str, Any]]
-):
+) -> TransactionMetricsSettings:
+
+    if killswitches.killswitch_matches_context(
+        "relay.drop-transaction-metrics", {"project_id": project.id}
+    ):
+        # Do not extract anything
+        return {"extractMetrics": [], "extractCustomTags": [], "customMeasurements": {"limit": 0}}
```
Comment (jjbayer) [inline src/sentry/relay/config/__init__.py:438]: This conditional is very similar to the feature flag check: https://github.com/getsentry/sentry/blob/1eca234240cfbc1b56216dba8bcee720064c59fc/src/sentry/relay/config/__init__.py#L443

But I guess it makes sense to have a kill switch mechanism independent of the feature flag?

File: src/sentry/runner/commands/killswitches.py [hunk 4]
```diff
@@ -110,6 +110,8 @@ def _push(killswitch_name, infile, yes):
             "Should the changes be applied?", default=False, show_default=True, abort=True
         )
     options.set(killswitch_name, new_option_value)
+    if callback := killswitches.ALL_KILLSWITCH_OPTIONS[killswitch_name].on_change:
+        callback(option_value, new_option_value)
```
Comment (jjbayer) [inline src/sentry/runner/commands/killswitches.py:114]: Maybe this could also be implemented as a signal handler on the `Option` model, but I personally prefer an independent trigger mechanism for kill switches.

File: src/sentry/options/defaults.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/options/defaults.py:366]: you need a separate PR for introducing this option, which we then have to deploy separately and set the option in prod to the default value

File: src/sentry/relay/config/__init__.py (hunk 3 above)
Comment (untitaker) [reply src/sentry/relay/config/__init__.py:438]: i think we should unify those checks tbh. i don't think it's a good idea to send more possible values that mean the same thing, because the potential is tehre that this doesnt parse correctly.

Comment (jjbayer) [reply src/sentry/relay/config/__init__.py:438]: I was hoping to prevent that by adding type hints to the return value, but I agree that a single check makes more sense, will refactor.

File: src/sentry/runner/commands/killswitches.py [hunk 5]
```diff
@@ -110,6 +110,8 @@ def _push(killswitch_name, infile, yes):
             "Should the changes be applied?", default=False, show_default=True, abort=True
         )
     options.set(killswitch_name, new_option_value)
+    if callback := killswitches.ALL_KILLSWITCH_OPTIONS[killswitch_name].on_change:
```
[the remark this answers is not in the document]
Comment (jjbayer) [reply src/sentry/runner/commands/killswitches.py:113]: Done.

**Comment by jjbayer** [comment]
Sorry for the many review requests, I merged the wrong branch.

File: requirements-dev.txt [hunk 6]
```diff
@@ -5,7 +5,7 @@ honcho>=1.0.0,<1.1.0
 mypy>=0.800,<0.900
 openapi-core==0.14.2
 pytest==6.1.0
-pytest-cov==2.11.1
+pytest-cov==3.0.0
```
Comment (andriisoldatenko) [inline requirements-dev.txt:8]: is it possible to move these changes to separate PR?

File: src/sentry/snuba/metrics/query.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 """ Classes needed to build a metrics query. Inspired by snuba_sdk.query. """
 import math
-from collections import Mapping
+from collections.abc import Mapping
```
Comment (andriisoldatenko) [inline src/sentry/snuba/metrics/query.py:3]: i think you have outdated base branch version, this change already in master https://github.com/getsentry/sentry/pull/36195

Comment (jjbayer) [reply src/sentry/snuba/metrics/query.py]: Thank you, yes, I accidentally merged master into this branch without updating the base branch.

File: src/sentry/killswitches.py [hunk 8]
```diff
@@ -20,10 +22,45 @@
...
+
+    old_project_ids = {ctx["project_id"] for ctx in old_option_value}
+    new_project_ids = {ctx["project_id"] for ctx in new_option_value}
+
+    # We want to recompute the project config for any project that was added
+    # or removed
+    changed_project_ids = old_project_ids ^ new_project_ids
+
+    with click.progressbar(changed_project_ids) as ids:
+        for project_id in ids:
```
Comment (untitaker) [inline src/sentry/killswitches.py:41]: please skip over `None` here, which can happen if the user wants to kill all projects

**Review by untitaker — APPROVED** [review]

**Review by jan-auer — APPROVED** [review]

File: src/sentry/options/defaults.py (hunk 2 above)
Comment (flub) [reply src/sentry/options/defaults.py]: Kind of off topic, but could we fix this so that we can happily introduce an option like this but it is just going to return the default (here the empty list) until the database thing is done?
